unit test index_rtree_exceptions_rst fails

388/388 Test #370: boost_geometry_index_rtree_exceptions_rst ...................***Failed    7.75 sec

```
$ ./index/test/rtree/exceptions/boost_geometry_index_rtree_exceptions_rst
Running 1 test case...
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:53: error: in "test_main_caller( argc_ argv )": check bgi::detail::rtree::utilities::are_counts_ok(tree, false) has failed

*** 7 failures are detected in the test module "Test Program"

```

If I change the code from
```
BOOST_CHECK(bgi::detail::rtree::utilities::are_counts_ok(tree, false));
```

into:
```
BOOST_CHECK_MESSAGE(bgi::detail::rtree::utilities::are_counts_ok(tree, false),
            "are_counts_ok failed"
            << " at i = " << i
            << " with max_calls = " << throwing_value::get_max_calls_ref());
```

I get the next output:
```
Running 1 test case...
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 34 with max_calls = 34
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 36 with max_calls = 36
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 38 with max_calls = 38
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 40 with max_calls = 40
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 42 with max_calls = 42
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 44 with max_calls = 44
.../boost/libs/geometry/index/test/rtree/exceptions/test_exceptions.hpp:56: error: in "test_main_caller( argc_ argv )": are_counts_ok failed at i = 46 with max_calls = 46

*** 7 failures are detected in the test module "Test Program"

```

If I comment the `BOOST_CHECK_THROW` line, the test passes.

I don't understand the test, and I don't understand the failure, so it's up to you...

--- a/include/boost/geometry/index/detail/rtree/rstar/insert.hpp
+++ b/include/boost/geometry/index/detail/rtree/rstar/insert.hpp
@@ -122,35 +122,59 @@ public:
         >::type sorted_elements_type;
 
         sorted_elements_type sorted_elements;
-        // If constructor is used instead of resize() MS implementation leaks here
-        sorted_elements.reserve(elements_count);                                                         // MAY THROW, STRONG (V, E: alloc, copy)
 
-        for ( typename elements_type::const_iterator it = elements.begin() ;
-              it != elements.end() ; ++it )
+        BOOST_TRY
         {
-            point_type element_center;
-            geometry::centroid(rtree::element_indexable(*it, translator), element_center,
-                               strategy);
-            sorted_elements.push_back(std::make_pair(
-                comparable_distance_pp::call(node_center, element_center, strategy),
-                *it));                                                                                  // MAY THROW (V, E: copy)
+            // If constructor is used instead of resize() MS implementation leaks here
+            sorted_elements.reserve(elements_count);                                                     // MAY THROW, STRONG (alloc)
+
+            for ( typename elements_type::const_iterator it = elements.begin() ;
+                  it != elements.end() ; ++it )
+            {
+                point_type element_center;
+                geometry::centroid(rtree::element_indexable(*it, translator), element_center,
+                                   strategy);
+                sorted_elements.push_back(std::make_pair(
+                    comparable_distance_pp::call(node_center, element_center, strategy),
+                    *it));                                                                              // MAY THROW (V, E: copy)
+            }
+
+            // sort elements by distances from center
+            std::partial_sort(
+                sorted_elements.begin(),
+                sorted_elements.begin() + reinserted_elements_count,
+                sorted_elements.end(),
+                distances_dsc<comparable_distance_type, element_type>);                                            // MAY THROW, BASIC (V, E: copy)
+
+            // copy elements which will be reinserted
+            result_elements.clear();
+            result_elements.reserve(reinserted_elements_count);                                         // MAY THROW, STRONG (V, E: alloc, copy)
+            for ( typename sorted_elements_type::const_iterator it = sorted_elements.begin() ;
+                  it != sorted_elements.begin() + reinserted_elements_count ; ++it )
+            {
+                result_elements.push_back(it->second);                                                  // MAY THROW (V, E: copy)
+            }
         }
-
-        // sort elements by distances from center
-        std::partial_sort(
-            sorted_elements.begin(),
-            sorted_elements.begin() + reinserted_elements_count,
-            sorted_elements.end(),
-            distances_dsc<comparable_distance_type, element_type>);                                                // MAY THROW, BASIC (V, E: copy)
-
-        // copy elements which will be reinserted
-        result_elements.clear();
-        result_elements.reserve(reinserted_elements_count);                                             // MAY THROW, STRONG (V, E: alloc, copy)
-        for ( typename sorted_elements_type::const_iterator it = sorted_elements.begin() ;
-              it != sorted_elements.begin() + reinserted_elements_count ; ++it )
+        BOOST_CATCH(...)
         {
-            result_elements.push_back(it->second);                                                      // MAY THROW (V, E: copy)
+            // NOTE: This code is here to prevent leaving the node in overflow state
+            //  (with more than max elements) after an exception is thrown during the
+            //  preparation of elements for reinsertion (#1399). At this point the node
+            //  has not been modified yet and still contains exactly max+1 elements
+            //  (one push_back triggered the overflow handling).
+            //  Remove the overflow element to restore the node invariant, matching
+            //  the behavior of the split exception handler.
+            result_elements.clear();
+
+            if (elements.size() > parameters.get_max_elements())
+            {
+                destroy_element<MembersHolder>::apply(elements.back(), allocators);
+                elements.pop_back();
+            }
+
+            BOOST_RETHROW
         }
+        BOOST_CATCH_END
 
         BOOST_TRY
         {

